// fis-kernel/lib/project.js
/*
 * fis
 * http://fis.baidu.com/
 */

'use strict';

exports.DEFAULT_REMOTE_REPOS = 'http://fis.baidu.com/repos';

exports.getSource = function(){
    var root = exports.getProjectPath(),
        source = {},
        //排除掉fis-conf文件和output文件夹
        project_exclude = new RegExp(
            '^' + fis.util.escapeReg(root + '/') +
                '(?:output\\b|fis-[^\\/]+$)',
            'i'),
        //可以自己定制排除
        include = fis.config.get('project.include'),
        exclude = fis.config.get('project.exclude');
    fis.util.find(root, null, project_exclude).forEach(function(file){
        //通过路径转话成File对象,存储了这个文件所有信息，是否发布，依赖文件，后缀等，还提供了setConent、getContent等方法
        file = fis.file(file);
        if(file.release && fis.util.filter(file.subpath, include, exclude)){
            source[file.subpath] = file;
        }
    });
    //以文件相对项目根目录为起始的文件地址(subpath)为key保存file对象列表
    return source;
};

//paths
var PROJECT_ROOT;
var TEMP_ROOT;

function getPath(root, args){
    if(args && args.length > 0){
        args = root + '/' + Array.prototype.join.call(args, '/');
        return fis.util(args);
    } else {
        return root;
    }
}

function initDir(path, title){
    if(fis.util.exists(path)){
        if(!fis.util.isDir(path)){
            fis.log.error('unable to set path[' + path + '] as ' + title + ' directory.');
        }
    } else {
        fis.util.mkdir(path);
    }
    path = fis.util.realpath(path);
    if(path){
        return path;
    } else {
        fis.log.error('unable to create dir [' + path + '] for ' + title + ' directory.');
    }
}

exports.getProjectPath = function(){
    if(PROJECT_ROOT){
        return getPath(PROJECT_ROOT, arguments);
    } else {
        fis.log.error('undefined project root');
    }
};

exports.setProjectRoot = function(path){
    if(fis.util.isDir(path)){
        PROJECT_ROOT = fis.util.realpath(path);
    } else {
        fis.log.error('invalid project root path [' + path + ']');
    }
};

exports.setTempRoot = function(tmp){
    TEMP_ROOT = initDir(tmp);
};

exports.getTempPath = function(){
    if(!TEMP_ROOT){
        var list = ['FIS_TEMP_DIR', 'LOCALAPPDATA', 'APPDATA', 'HOME'];
        var name = fis.cli && fis.cli.name ? fis.cli.name : 'fis';
        var tmp;
        for(var i = 0, len = list.length; i < len; i++){
            if(tmp = process.env[list[i]]){
                break;
            }
        }
        tmp = tmp || __dirname + '/../';
        exports.setTempRoot(tmp + '/.' + name + '-tmp');
    }
    return getPath(TEMP_ROOT, arguments);
};

exports.getCachePath = function(){
    return getPath(exports.getTempPath('cache'), arguments);
};
