// fis-kernel/lib/release.js
/*
 * fis
 * http://fis.baidu.com/
 */

'use strict';

/**
 * @type {Function}
 * @param {Object} [opt]
 * @param {Function} callback
 * opt = {
 *     include : {RegExp} find include filter,
 *     exclude : {RegExp} find exclude filter,
 *     debug    : {Boolean} debug model,
 *     optimize : {Boolean} if optimize,
 *     pack     : {Boolean|String} if package,
 *     lint     : {Boolean} if lint,
 *     test     : {Boolean} if test,
 *     hash     : {Boolean} if with hash,
 *     domain   : {Boolean} if with domain,
 *     beforeEach : {Function} before compile each file callback,
 *     afterEach : {Function} after compile each file callback,
 *     beforePack : {Function} before pack file callback,
 *     afterPack : {Function} after pack file callback
 * }
 */

var exports = module.exports = function(opt, callback){
    if(typeof opt === 'function'){
        callback = opt;
        opt = {};
    } else {
        opt = opt || {};
    }
    //构建文件系统
    var ret = {
        //将项目文件夹下所有的文件都初始花成一个个File对象
        //源码在file.js里面
        src : fis.project.getSource(),
        ids : {},
        pkg : {},
        //map资源表
        map : {
            res : {},
            pkg : {}
        }
    };
    opt.hash = opt.md5 > 0;
    fis.compile.setup(opt);
    //遍历所有的文件进行编译+打包处理
    fis.util.map(ret.src, function(subpath, file){
        //回掉函数
        if(opt.beforeEach) {
            opt.beforeEach(file, ret);
        }
        //单文件编译过程，流程图图中左侧部分，配置时会和后缀明有关
        file = fis.compile(file);
        if(opt.afterEach) {
            opt.afterEach(file, ret);
        }
        //这里添加制作map依赖表
        if(file.release && file.useMap) {
            //add resource map
            var id = file.getId();
            ret.ids[id] = file;
            //这里自动添加相同路径下同名的文件
            //js会添加css
            if(file.isJsLike){
                file.addSameNameRequire('.css');
            //html会添加js+css
            //note:这里就是实现{%widge%}t的基础了
            } else if(file.isHtmlLike){
                file.addSameNameRequire('.js');
                file.addSameNameRequire('.css');
            }
            //这里是制作依赖关系map
            var res = ret.map.res[id] = {
                uri  : file.getUrl(opt.hash, opt.domain),
                type : file.rExt.replace(/^\./, '')
            };
            for(var key in file.extras){
                if(file.extras.hasOwnProperty(key)){
                    res.extras = file.extras;
                    break;
                }
            }
            if(file.requires && file.requires.length){
                res.deps = file.requires;
            }
        }
    });
    
    //project root
    var root = fis.project.getProjectPath();
    
    var ns = fis.config.get('namespace');
   //以下是流程图右侧部分
   //都和文件后缀名称没有关系，所以在配置中无需加后缀名称
    //get pack config
    var conf = fis.config.get('pack');
    if(typeof conf === 'undefined'){
        //from fis-pack.json
        var file = root + '/fis-pack.json';
        if(fis.util.isFile(file)){
            fis.config.set('pack', conf = fis.util.readJSON(file));
        }
    }
    
    //package callback
    var cb = function(packager, settings, key){
        fis.log.debug('[' + key + '] start');
        //执行
        packager(ret, conf, settings, opt);
        fis.log.debug('[' + key + '] end');
    };

    //prepackage
    fis.util.pipe('prepackager', cb, opt.prepackager);
    
    //package
    if(opt.pack){
        //package
        fis.util.pipe('packager', cb, opt.packager);
        //css sprites
        fis.util.pipe('spriter', cb, opt.spriter);
    }
    
    //postpackage
    fis.util.pipe('postpackager', cb, opt.postpackager);
    
    //create map.json
    var map = fis.file(root, (ns ? ns + '-' : '') + 'map.json');
    if(map.release){
        //将map压缩
        map.setContent(JSON.stringify(ret.map, null, opt.optimize ? null : 4));
        //放到
        ret.pkg[map.subpath] = map;
    }
    
    //done
    if(callback){
        callback(ret);
    }
};
