// fis-command-release/release.js
/*
 * fis
 * http://fis.baidu.com/
 */

'use strict';

exports.name = 'release';
exports.desc = 'build and deploy your project';
//这个模块只提供了一个`regisiter`方法供调用
exports.register = function(commander){//参数是commander模块
    //watch监听文件修改功能
    function watch(opt){
        var root = fis.project.getProjectPath();
        var timer = -1;
        var safePathReg = /[\\\/][_\-.\s\w]+$/i;
        function listener(path){
            if(safePathReg.test(path)){
                clearTimeout(timer);
                timer = setTimeout(function(){
                    release(opt);
                }, 500);
            }
        }
        //这是个监听文件修改的模块库-->https://github.com/paulmillr/chokidar
        require('chokidar')
            .watch(root, {
                //忽略output文件夹和fis-conf.js和当前(.)文件夹
                ignored : /[\/\\](?:output\b[^\/\\]*([\/\\]|$)|\.|fis-conf\.js$)/i,
                // usePolling: false,
                persistent: true
            })
            //文件新增时触发
            .on('add', listener)
            //文件修改时触发
            .on('change', listener)
            //文件删除时触发
            .on('unlink', listener)
            .on('error', function(err){
                //fis.log.error(err);
            });
    }
     //输出时间的方法
    function time(fn){
        process.stdout.write('\n δ '.bold.yellow);
        var now = Date.now();
        fn();
        process.stdout.write((Date.now() - now + 'ms').green.bold);
        process.stdout.write('\n');
    }
   //reload方法
    var LRServer, LRTimer;
    function reload(){
        if(LRServer && LRServer.connections) {
            fis.util.map(LRServer.connections, function(id, connection){
                try {
                    connection.send({
                        command: 'reload',
                        path: '*',
                        liveCSS: true
                    });
                } catch (e) {
                    try {
                        connection.close();
                    } catch (e) {}
                    delete LRServer.connections[id];
                }
            });
        }
    }
    
    var lastModified = {};
    var collection = {};
    var deploy = require('./lib/deploy.js');
    
    deploy.done = function(){
        clearTimeout(LRTimer);
        LRTimer = setTimeout(reload, 200);
    };
    
    function release(opt){
        var flag, cost, start = Date.now();
        //这句其实就是console.log功能
        process.stdout.write('\n Ω '.green.bold);
        //循环处理所有文件前处理前调用的方法
        opt.beforeEach = function(){
            flag = opt.verbose ? '' : '.';
            cost = (new Date).getTime();
        };
        //循环处理所有文件后处理前调用的方法
        opt.afterEach = function(file){
            //cal compile time
            cost = (new Date).getTime() - cost;
            if(cost > 200){
                flag = flag.bold.yellow;
                fis.log.debug(file.realpath);
            } else if(cost < 100){
                flag = flag.grey;
            }
            //文件最后修改时间
            var mtime = file.getMtime().getTime();
            //collect file to deploy
            //在配置文件中匹配到的release
            if(file.release && lastModified[file.subpath] !== mtime){
                if(!collection[file.subpath]){
                    process.stdout.write(flag);
                }
                lastModified[file.subpath] = mtime;
                collection[file.subpath] = file;
            }
        };

        //单文件编译前处理前调用的方法
        opt.beforeCompile = function(file){
            collection[file.subpath] = file;
            process.stdout.write(flag);
        };
        
        try {
            //release
            //调用fis-kernel里面的release方法
            fis.release(opt, function(ret){
                process.stdout.write(
                    (opt.verbose ? '' : ' ') +
                    (Date.now() - start + 'ms').bold.green + '\n'
                );

                //遍历项目里面的所有文件
                for(var item in collection){
                    if(collection.hasOwnProperty(item)){
                        if(opt.unique){
                            time(fis.compile.clean);
                        }
                        deploy(opt, collection);
                        deploy(opt, ret.pkg);
                        collection = {};
                        return;
                    }
                }
            });
        } catch(e) {
            process.stdout.write('\n [ERROR] ' + (e.message || e) + '\n');
            if(opt.watch){
                process.stdout.write('\u0007');
            } else if(opt.verbose) {
                throw e;
            } else {
                process.exit(1);
            }
        }
    }
    //release明显会注册这么多种格式的命令
    commander
        //参数第一个是短命令和长命令，以及值
        //参数第二个描述命令的功能(description)，在用户输出帮助时用得到
        //参数第三个会被强制转成成的类型
        //参数第五个是默认值
        .option('-d, --dest <names>', 'release output destination', String, 'preview')
        .option('-m, --md5 [level]', 'md5 release option', Number)
        .option('-D, --domains', 'add domain name', Boolean, false)
        .option('-l, --lint', 'with lint', Boolean, false)
        .option('-t, --test', 'with unit testing', Boolean, false)
        .option('-o, --optimize', 'with optimizing', Boolean, false)
        .option('-p, --pack', 'with package', Boolean, true)
        .option('-w, --watch', 'monitor the changes of project')
        .option('-L, --live', 'automatically reload your browser')
        .option('-c, --clean', 'clean compile cache', Boolean, false)
        .option('-r, --root <path>', 'set project root')
        .option('-f, --file <filename>', 'set fis-conf file')
        .option('-u, --unique', 'use unique compile caching', Boolean, false)
        .option('--verbose', 'enable verbose output', Boolean, false)
        //action是注册所有命令的回调函数
        .action(function(){
            
            var options = arguments[arguments.length - 1];
            
            fis.log.throw = true;
           //定义信息输出级别
            //configure log
            if(options.verbose){
                fis.log.level = fis.log.L_ALL;
            }
            var root, conf, filename = 'fis-conf.js';
            //例如：fis release -f fis.conan.js
            //此事option.file等于 fis.conan.js
            if(options.file){
                //判断是不是文件
                if(fis.util.isFile(options.file)){
                    //返回解析后的路径
                    conf = fis.util.realpath(options.file);
                } else {
                    fis.log.error('invalid fis config file path [' + options.file + ']');
                }
            }
            //项目根目录
            if(options.root){
                root = fis.util.realpath(options.root);
                if(fis.util.isDir(root)){
                    if(!conf && fis.util.isFile(root + '/' + filename)){
                        conf = root + '/' + filename;
                    }
                    delete options.root;
                } else {
                    fis.log.error('invalid project root path [' + options.root + ']');
                }
            } else {
                //默认的根目录使用当前目录
                root = fis.util.realpath(process.cwd());
                if(!conf){
                    //try to find fis-conf.js
                    var cwd = root, pos = cwd.length;
                    do {
                        cwd  = cwd.substring(0, pos);
                        conf = cwd + '/' + filename;
                        if(fis.util.exists(conf)){
                            root = cwd;
                            break;
                        } else {
                            conf = false;
                            pos = cwd.lastIndexOf('/');
                        }
                    } while(pos > 0);
                }
            }
           //设置项目目录
            //init project
            fis.project.setProjectRoot(root);
           //设置流程名称，在*uix里面可以用ps显示出来
            process.title = 'fis ' + process.argv.splice(2).join(' ') + ' [ ' + root + ' ]';
           //调用执行配置文件(fis-conf.js)
            if(conf){
                var cache = fis.cache(conf, 'conf');
                if(!cache.revert()){
                    options.clean = true;
                    cache.save();
                }
                require(conf);
            } else {
                fis.log.warning('missing config file [' + filename + ']');
            }
           //例如：fis release -c是这里是true
            if(options.clean){
                time(function(){
                    fis.cache.clean('compile');
                });
            }
            delete options.clean;
            
            //domain, fuck EventEmitter
            if(options.domains){
                options.domain = true;
                delete options.domains;
            }
            //自动刷新浏览器功能
            if(options.live){
                var LiveReloadServer = require('livereload-server-spec');
                var port = fis.config.get('livereload.port', 8132);
                LRServer = new LiveReloadServer({
                    id: 'com.baidu.fis',
                    name: 'fis-reload',
                    version : fis.cli.info.version,
                    port : port,
                    protocols: {
                        monitoring: 7
                    }
                });
                LRServer.on('livereload.js', function(req, res) {
                    var script = fis.util.fs.readFileSync(__dirname + '/vendor/livereload.js');
                    res.writeHead(200, {
                        'Content-Length': script.length,
                        'Content-Type': 'text/javascript',
                        'Connection': 'close'
                    });
                    res.end(script);
                });
                LRServer.listen(function(err) {
                    if (err) {
                        err.message = 'LiveReload server Listening failed: ' + err.message;
                        fis.log.error(err);
                    }
                });
                process.stdout.write('\n Ψ '.bold.yellow + port + '\n');
                //fix mac livereload
                process.on('uncaughtException', function (err) {
                    if(err.message !== 'read ECONNRESET') throw  err;
                });
                //delete options.live;
            }
           //md5时间戳功能设置
            switch (typeof options.md5){
                case 'undefined':
                    options.md5 = 0;
                    break;
                case 'boolean':
                    options.md5 = options.md5 ? 1 : 0;
                    break;
                default :
                    options.md5 = isNaN(options.md5) ? 0 : parseInt(options.md5);
            }
            //md5 > 0, force release hash file
            options.hash = options.md5 > 0;
           //如果监听(watch)
            if(options.watch){
                watch(options);
            } else {
                //开始发布
                release(options);
            }
        });
};
